### ISSUE TITLE
cntk for WGAN-GP

### ISSUE THREAD

**Issue by DabiaoMa** [Microsoft/CNTK#2140]
I am trying to implement WGAN-GP with cntk, but here comes a big difficulty.

Suppose x is an input variable and  F is the discriminator neural network, then first we need to get the gradient over x : ∇x(F) w.r.t. parameters of F.
And then, we need to constrain the gradient ∇x(F), that is, make this term (||∇x(F)|| - 1) a loss.

Would you please tell me how to implement this loss??

Best regards.

**Comment by ebarsoumMS** [comment]
Sadly, CNTK doesn't support symbolic gradient so you can't use gradient as part of the loss (It is in our backlog). A workaround is to use F(x+r) - F(x), where r is a small uniform random value. You can use CNTK.random to generate random values.

**Comment by n17s** [comment]
Right, the workaround for now is basically that F(x+r) for small r is F(x)+r'*∇F so an estimator like
(F(x+ε*r) - F(x-ε*r))/(2*ε) computes r'*∇F and ((F(x+ε*r) - F(x-ε*r))/(2*ε))^2 computes ∇F'*r*r'*∇F whose expected value is ||∇F||^2 if r is drawn from a standard normal distribution.

Obviously this workaround can be a noisy estimate, and we have not verified that it actually makes GAN training better.

**Comment by haixpham** [comment]
May I know when will this feature be added to CNTK?

**Comment by 1070933148** [comment]
excuse me, does cntk support symbolic gradient today? and how to use Function.gradients, I could not find description in API doc.

**Comment by ebarsoumMS** [comment]
Currently, we don't have it. It is in our backlog, but I don't have an ETA for it.

**Comment by 1070933148** [comment]
Thanks for your reply, but could you take a look at #2589 .

**Comment by 1070933148** [comment]
@ebarsoumMS Thanks for your reply, but could you take a look at #2589 .
